67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,1088 @@
+
+			cmdbufSubmitAndWait(cmdbufBlas, getComputeQueue(), 39);
+
+// TODO: SUBMIT THE BLAS BUILD FIRST AND AWAIT ITS COMPLETION WITH A SEMAPHORE SIGNAL
+
+			auto cmdbufTlas = getSingleUseCommandBufferAndBegin(pool);
+
+			//size_t asSizes[1];
+			//m_device->getQueryPoolResults(queryPool.get(), 0, queryCount, asSizes, sizeof(size_t), IQueryPool::WAIT_BIT);
+			//
+			//auto cleanupBlas = gpuBlas;
+			//{
+			//	IGPUBuffer::SCreationParams params;
+			//	params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_ACCELERATION_STRUCTURE_STORAGE_BIT;
+			//	params.size = asSizes[0];
+			//	smart_refctd_ptr<IGPUBuffer> asBuffer = createBuffer(params);
+
+			//	IGPUBottomLevelAccelerationStructure::SCreationParams blasParams;
+			//	blasParams.bufferRange.buffer = asBuffer;
+			//	blasParams.bufferRange.offset = 0u;
+			//	blasParams.bufferRange.size = asSizes[0];
+			//	blasParams.flags = IGPUBottomLevelAccelerationStructure::SCreationParams::FLAGS::NONE;
+			//	gpuBlas = m_device->createBottomLevelAccelerationStructure(std::move(blasParams));
+			//}
+
+			//IGPUBottomLevelAccelerationStructure::CopyInfo copyInfo;
+			//copyInfo.src = cleanupBlas.get();
+			//copyInfo.dst = gpuBlas.get();
+			//copyInfo.mode = IGPUBottomLevelAccelerationStructure::COPY_MODE::COMPACT;
+			//cmdbuf->copyAccelerationStructure(copyInfo);
+
+			// build top level AS
+			{
+				const uint32_t instancesCount = 1;	// TODO: temporary for now
+				IGPUTopLevelAccelerationStructure::DeviceInstance instances[instancesCount];
+				core::matrix3x4SIMD identity;
+				instances[0].blas.deviceAddress = gpuBlas->getReferenceForDeviceOperations().deviceAddress;
+
+				{
+					size_t bufSize = sizeof(IGPUTopLevelAccelerationStructure::DeviceStaticInstance);
+					IGPUBuffer::SCreationParams params;
+					params.usage = bitflag(IGPUBuffer::EUF_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT) | IGPUBuffer::EUF_STORAGE_BUFFER_BIT |
+						IGPUBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF | IGPUBuffer::EUF_TRANSFER_DST_BIT | IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+					params.size = bufSize;
+					instancesBuffer = createBuffer(params);	// does this need host visible memory?
+
+					SBufferRange<IGPUBuffer> range = { .offset = 0u, .size = bufSize, .buffer = instancesBuffer };
+					cmdbufTlas->updateBuffer(range, instances);
+				}
+
+				// make sure instances upload complete first
+				{
+					SMemoryBarrier memBarrier;
+					memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS;
+					memBarrier.srcAccessMask = ACCESS_FLAGS::TRANSFER_WRITE_BIT;
+					memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::ACCELERATION_STRUCTURE_BUILD_BIT;
+					memBarrier.dstAccessMask = ACCESS_FLAGS::ACCELERATION_STRUCTURE_WRITE_BIT;
+					cmdbufTlas->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, { .memBarriers = {&memBarrier, 1} });
+				}
```

Comment:
this also should contain and source flags for the BLAS compaction

Reply:
hmm technically BLAS compaction counts as a TRANSFER because its a copy. so flags all good